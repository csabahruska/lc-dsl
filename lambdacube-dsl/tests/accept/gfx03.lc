main = let
  blendFun x = Blend x ((SrcAlpha,OneMinusSrcAlpha),(SrcAlpha,OneMinusSrcAlpha)) (V4F 1.0 1.0 1.0 1.0)
  blend'' = blendFun (FuncAdd,FuncAdd)
  blend = Blend (FuncAdd,FuncAdd) ((SrcAlpha,OneMinusSrcAlpha),(SrcAlpha,OneMinusSrcAlpha)) (V4F 1.0 1.0 1.0 1.0)
  blend' = NoBlending
  bgColor = V4F 0.5 0.0 0.4 1.0
  bgColor' = V4F 0.2 0.2 0.4 1.0
  emptyFB           = FrameBuffer (DepthImage #1 1000.0,ColorImage #1 bgColor)
  polyMode          = PolygonLine 20.0
  polyMode'         = PolygonFill
  polyMode''        = PolygonPoint (PointSize 10.0)
  cull = CullNone
  cull' = CullFront CW
  rasterCtx         = TriangleCtx cull polyMode' NoOffset LastVertex
  fragmentCtx       = AccumulationContext (DepthOp Less False, ColorOp blend (V4B True True True True))
  modelViewProj     = Uni (IM44F "MVP")
  vertexShader v    = VertexOut (PrimMulMatVec modelViewProj v) 1.0 () (Flat v)
  vertexStream      = Fetch "stream" Triangles (IV4F "position")
  primitiveStream   = Transform vertexShader vertexStream
  fragmentStream    = Rasterize rasterCtx primitiveStream
  fragmentShader' v = FragmentOutRastDepth (Const (V4F 1.0 0.4 0.0 0.2))
  fragmentShader v  = FragmentOutRastDepth (PrimMul v (Const (V4F 1.0 1.4 1.0 0.6)))
  frame             = Accumulate fragmentCtx PassAll fragmentShader fragmentStream emptyFB
 in ScreenOut frame
