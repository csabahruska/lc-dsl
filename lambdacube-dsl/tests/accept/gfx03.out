Pipeline
  { textures = []
  , samplers = []
  , targets =
      [ RenderTarget
          { renderTargets =
              [ ( Depth , Just (Framebuffer Depth) )
              , ( Color , Just (Framebuffer Color) )
              ]
          }
      ]
  , programs =
      [ Program
          { programUniforms = fromList [ ( "MVP2" , M44F ) ]
          , programStreams = fromList [ ( "v" , ( "position" , V3F ) ) ]
          , programInTextures = fromList []
          , programOutput = [ ( "f0" , V4F ) ]
          , vertexShader =
              "#version 330 core\nuniform mat4 MVP2 ;\nin vec3 v ;\nvoid main() {\ngl_Position = ( MVP2 ) * ( vec4( v ,1.0) );\ngl_PointSize = 1.0;\n}\n"
          , geometryShader = Nothing
          , fragmentShader =
              "#version 330 core\nout vec4 f0 ;\nvoid main() {\nf0 = vec4 ( 0.0,0.4,0.0,1.0 );\n}\n"
          }
      , Program
          { programUniforms = fromList [ ( "MVP2" , M44F ) ]
          , programStreams = fromList [ ( "v" , ( "position" , V3F ) ) ]
          , programInTextures = fromList []
          , programOutput = [ ( "f0" , V4F ) ]
          , vertexShader =
              "#version 330 core\nuniform mat4 MVP2 ;\nin vec3 v ;\nsmooth out vec4 v0 ;\nvoid main() {\nv0 = vec4( v ,1.0);\ngl_Position = ( MVP2 ) * ( vec4( v ,1.0) );\ngl_PointSize = 1.0;\n}\n"
          , geometryShader = Nothing
          , fragmentShader =
              "#version 330 core\nsmooth in vec4 v0 ;\nout vec4 f0 ;\nvoid main() {\nf0 = ( v0 ) + ( vec4 ( 1.0,1.4,1.0,0.6 ) );\n}\n"
          }
      , Program
          { programUniforms = fromList [ ( "MVP" , M44F ) ]
          , programStreams = fromList [ ( "v" , ( "position4" , V4F ) ) ]
          , programInTextures = fromList []
          , programOutput = [ ( "f0" , V4F ) ]
          , vertexShader =
              "#version 330 core\nuniform mat4 MVP ;\nin vec4 v ;\nflat out vec4 v0 ;\nvoid main() {\nv0 = v;\ngl_Position = ( MVP ) * ( v );\ngl_PointSize = 1.0;\n}\n"
          , geometryShader = Nothing
          , fragmentShader =
              "#version 330 core\nflat in vec4 v0 ;\nout vec4 f0 ;\nvoid main() {\nf0 = ( v0 ) * ( vec4 ( 1.0,1.4,1.0,0.6 ) );\n}\n"
          }
      ]
  , slots =
      [ Slot
          { slotName = "stream"
          , slotUniforms = fromList [ ( "MVP2" , M44F ) ]
          , slotStreams = fromList [ ( "position" , V3F ) ]
          , slotPrimitive = Triangles
          , slotPrograms = [ 0 , 1 ]
          }
      , Slot
          { slotName = "stream4"
          , slotUniforms = fromList [ ( "MVP" , M44F ) ]
          , slotStreams = fromList [ ( "position4" , V4F ) ]
          , slotPrimitive = Triangles
          , slotPrograms = [ 2 ]
          }
      ]
  , commands =
      [ SetRenderTarget 0
      , ClearRenderTarget
          [ ( Depth , VFloat 1000.0 )
          , ( Color , VV4F (V4 0.5 0.0 0.4 1.0) )
          ]
      , SetRasterContext
          (TriangleCtx CullNone PolygonFill NoOffset LastVertex)
      , SetAccumulationContext
          AccumulationContext
            { accViewportName = Nothing
            , accOperations =
                [ DepthOp Less False
                , ColorOp
                    (Blend
                       ( FuncAdd , FuncAdd )
                       ( ( SrcAlpha , OneMinusSrcAlpha )
                       , ( SrcAlpha , OneMinusSrcAlpha )
                       )
                       (V4 1.0 1.0 1.0 1.0))
                    (VV4B (V4 True True True True))
                ]
            }
      , SetProgram 2
      , RenderSlot 1
      , SetRasterContext
          (TriangleCtx CullNone PolygonFill NoOffset FirstVertex)
      , SetAccumulationContext
          AccumulationContext
            { accViewportName = Nothing
            , accOperations =
                [ DepthOp Less False
                , ColorOp NoBlending (VV4B (V4 True True False False))
                ]
            }
      , SetProgram 1
      , RenderSlot 0
      , SetRasterContext
          (TriangleCtx CullNone (PolygonLine 20.0) NoOffset FirstVertex)
      , SetAccumulationContext
          AccumulationContext
            { accViewportName = Nothing
            , accOperations =
                [ DepthOp Always False
                , ColorOp NoBlending (VV4B (V4 True True False False))
                ]
            }
      , SetProgram 0
      , RenderSlot 0
      ]
  }